### ISSUE TITLE
Disabling "smart formatting" of HTML class attribute

### ISSUE THREAD

**Issue by luukdv** [prettier/prettier#10918]
_"We get a lot of requests for adding options, but Prettier is built on the principle of being opinionated about code formatting."_

I totally understand this, and so I apologize beforehand for opening this issue. The reason I think this is a valid issue is because the opinion about this seems to have _changed_ from the perspective of Prettier, even though it was mentioned that there's no community interest for this change (and a lot of pushback) in the [introducing pull request](https://github.com/prettier/prettier/pull/7865).

**The issue**
When using frameworks like [Tailwind](https://github.com/tailwindlabs/tailwindcss), readability has lowered drastically since this feature:

<img width="320" src="https://user-images.githubusercontent.com/1065835/117627441-13b4c680-b170-11eb-824b-d22f07456e84.png" />

**References**
- [#836345253](https://github.com/prettier/prettier/pull/7865#issuecomment-836345253)
- [#836536543](https://github.com/prettier/prettier/pull/7865#issuecomment-836536543)
- [#836652940](https://github.com/prettier/prettier/pull/7865#issuecomment-836652940)
- [#839911172](https://github.com/prettier/prettier/pull/7865#issuecomment-839911172)
- [#841662012](https://github.com/prettier/prettier/pull/7865#issuecomment-841662012)

I'd love to still be able to use Prettier, but this is a blocking issue for me and many others.

**The solution**
Respecting the `class` attribute to stay on a single line, and only applying formatting it when a user already has a multi-line approach for `class`. Alternatively, an option to disable this feature.

Thanks in advance for considering this!

**Comment by thorn0** [comment]
> Respecting the class attribute to stay on a single line, and only applying formatting it when a user already has a multi-line approach for class. 

This sounds like a good trade-off to me, but probably it can be improved. I guess developers who don't use Tailwind and have longer class names would prefer to keep the new behavior. So can Prettier probably somehow detect whether a single-line `class` attribute is "Tailwind-like"?

**Comment by alexander-akait** [comment]
@luukdv 

> Respecting the class attribute to stay on a single line, and only applying formatting it when a user already has a multi-line approach for class. Alternatively, an option to disable this feature.

just interesting, what is the problem with multi line, I found it more readable

**Comment by tkalmar** [comment]
I find the single line approach problematic when having many classes (like in a tailwind app with multiple breakpoints)
The big advantage i see is diffability (is this even a word?) i see clearly which classes are changed when i look at the diff.
There are multiple approaches i can imagine:
* don't switch style either single line / multiline  warn if mixed at most
* leave class as is. we use multiline approach to group matching classes for different breakpoints, some group by prefix ... it is very hard to find a OOB solution which fit's all needs
* make the grouping configurable/pluggable so i can enforce multiline but decide which classes are kept together (the current solution with a little bit more flexibility

**Comment by thorn0** [comment]
> it is very hard to find a OOB solution which fit's all needs

Nonetheless, I'd say we're almost there because so far looks like the current OOB solution is good enough for everyone else except Tailwind users. So we only need to find a way to improve it for Tailwind.

@luukdv Could you please replace the image in your post with a code block, so that it would be easy to experiment with reformatting that code?

**Comment by thorn0** [comment]
Just another brain storm comment, how about grouping all ungrouped classes at the end of the class list, after all the prefixed groups (only if there are prefixed groups)?

```
foo
bar
baz-one baz-two baz-three
qux-one qux-two
quuux lorem-ipsum bla-bla
```

**Comment by luukdv** [comment]
> I found it more readable
@alexander-akait I guess it's a matter of preference then. For me, if I consider the following:

```
[...]

<div class="grid grid-cols-2 sm:grid-cols-3 gap-2 sm:gap-4 md:gap-8 max-w-screen-xl mx-auto px-5 sm:px-10">
  <a href="" class="relative bg-gray-200 group block transform border-2 border-white rounded-lg">
    <img src="" alt="" class="z-10 object-cover absolute top-0 left-0 w-full h-full block h-auto w-full" />

    <div class="absolute z-30 overflow-hidden hidden lg:flex flex-col font-medium justify-center text-white px-8 top-0 left-0 w-full h-full transition-opacity ease-cubic duration-400 opacity-0 lg:group-hover:opacity-100">
      <p class="mb-4 last:mb-0 text-gray-800 italic inline-block text-sm px-2 leading-relaxed z-10">
        Some text
      </p>
    </div>
  </a>
</div>

[...]
```

I'd see what's going on HTML-wise in a glance. With the new approach, it's less readable, especially since there's even more of these blocks above and below:

```
[...]

<div
  class="
    grid
    grid-cols-2
    sm:grid-cols-3
    gap-2
    sm:gap-4
    md:gap-8
    max-w-screen-xl
    mx-auto
    px-5
    sm:px-10
  "
>
  <a
    href=""
    class="
      relative
      bg-gray-200
      group
      block
      transform
      border-2
      border-white
      rounded-lg
    "
  >
    <img
      src=""
      alt=""
      class="
        z-10
        object-cover
        absolute
        top-0
        left-0
        w-full
        h-full
        block
        h-auto
        w-full
      "
    />

    <div
      class="
        absolute
        z-30
        overflow-hidden
        hidden
        lg:flex
        flex-col
        font-medium
        justify-center
        text-white
        px-8
        top-0
        left-0
        w-full
        h-full
        transition-opacity
        ease-cubic
        duration-400
        opacity-0
        lg:group-hover:opacity-100
      "
    >
      <p
        class="
          mb-4
          last:mb-0
          text-gray-800
          italic
          inline-block
          text-sm
          px-2
          leading-relaxed
          z-10
        "
      >
        Some text
      </p>
    </div>
  </a>
</div>

[...]
```

> Could you please replace the image in your post with a code block, so that it would be easier to experiment with reformatting that code?
@thorn0 Sure thing:

```
<div
  class="
    rounded-t
    lg:rounded
    overflow-hidden
    border
    border-gray-200
    group-hover:border-transparent
    group-hover:ring-2
    ring-offset-2
    ring-secondary-500
    object-cover
    w-full
    h-full
    absolute
    inset-0
    cursor-pointer
    z-1
  "
/>

<div
  class="
    bg-secondary-500
    text-white
    cursor-pointer
    hidden
    lg:block
    lg:invisible
    lg:group-hover:visible
    rounded
    shadow-lg
    px-6
    py-2
    font-medium
    absolute
    position-center
    z-3
  "
>
  <span>View</span>
</div>

<div
  class="
    lg:hidden
    bg-white
    bg-opacity-90
    rounded
    shadow-lg
    px-2
    py-1
    text-sm
    font-medium
    absolute
    right-[5px]
    top-[5px]
  "
/>
```

**Comment by alexander-akait** [comment]
@thorn0 looks reasonable to have this in one line

**Comment by vinayakkulkarni** [comment]
Firstly, thanks for improving DX by bringing such brilliant features in prettier! Cheers!

Right now, I'm facing issues while saving the code.. <kbd>⌘</kbd> + <kbd>S</kbd> causes the linted file to revert the css classes, have to again run the linter manually to save the styles as per the new prettier "smart formatting" rules. Demo:

https://user-images.githubusercontent.com/19776877/119192055-1e8b2780-ba9d-11eb-9a2c-7cb0b371733a.mov

There could/should be a way one can turn this feature on or make it work with the eslint-plugin

**Comment by Shahaed** [comment]
> > I found it more readable
> 
> @alexander-akait I guess it's a matter of preference then. For me, if I consider the following:
> 
> ```
> [...]
> 
> <div class="absolute z-30 overflow-hidden hidden lg:flex flex-col font-medium justify-center text-white px-8 top-0 left-0 w-full h-full transition-opacity ease-cubic duration-400 opacity-0 lg:group-hover:opacity-100">
>   <a href="" class="relative bg-gray-200 group block transform border-2 border-white rounded-lg">
>   </a>
> </div>
> 
> [...]
> ```
> 

> 
> ```
> [...]
> 
> <div
>   class="
>     grid
>     grid-cols-2
>     sm:grid-cols-3
>     gap-2
>     sm:gap-4
>     md:gap-8
>     max-w-screen-xl
>     mx-auto
>     px-5
>     sm:px-10
>   "
> 
> [...]
> ```

I would say the HTML is much easier to read in the first example while the css is much easier to read in the next. I think this a tough problem for many Tailwind users (myself included). 

[Netlify solution was to go multiline css](https://www.netlify.com/blog/2021/03/23/from-semantic-css-to-tailwind-refactoring-the-netlify-ui-codebase/#readability) so I don't think all tailwind users agree this "smart formatting" is "bad". Tough problem however.

My proposed solution would be what Prettier has already implemented on multi lined objects: https://prettier.io/docs/en/rationale.html#multi-line-objects

I believe this would keep everyone happy?

**Comment by sabberworm** [comment]
I, too, am blocked from upgrading to 2.3.0 because of #7865. We use [HTL](https://github.com/adobe/htl-spec) (formerly Sightly), a templating language based on HTML. Unfortunately we use quite a few expressions in the `class` attribute, e.g.:

```html
<div class="cmp-accordion__button${item.name in accordion.expandedItems || accordion.extendedItems.size == 1 ? ' cmp-accordion__button--expanded' : ''}"></div>
```

which prettier 2.3 will reformat as

```html
<div
	class="
		cmp-accordion__button${item.name
		in
		accordion.expandedItems
		||
		accordion.extendedItems.size
		==
		1
		?
		'
		cmp-accordion__button--expanded'
		:
		''}
	"><div>

**Comment by adamwathan** [comment]
Just wanted to share my two cents from the Tailwind team — I think there are definitely interesting things that could be done formatting-wise but I worry it's going to be hard to do within Prettier itself because I think the best solutions require more semantic knowledge of the classes being used.

I can say my personal preference would be that Prettier did nothing with the class attribute, but if it's going to do anything I would really prefer if it _doesn't_ do anything to the order of the classes. Lots of folks in the community use tools like [Headwind](https://github.com/ryanhhhh/headwind) to sort their classes for them in a Tailwind-aware way, and we really want to build functionality like this into our own editor extension as well in the near future.

Personally I've always been very satisfied with just enabling word-wrap in my editor, but totally get that others like to have things split up.

Is this a big issue for people outside Tailwind projects? If not, maybe we could prioritize working on a Tailwind plugin for Prettier and make that the solution vs. forcing Prettier to care too much about Tailwind's naming conventions and stuff?

If there are any questions I can answer or input I can provide that would be helpful please let me know!  Absolutely _love_ Prettier and really appreciate the work of everyone that contributes to this project. Especially grateful that people are taking it upon themselves to try and come up with solutions to stuff like this long class attribute problem and that people are taking Tailwind into consideration when working on it ❤️

**Comment by royeden** [comment]
Hi!

Adding a comment regarding the impact of this prettier feature inside of Vue 2 (using laravel and laravel mix, with `"laravel-mix": "^6.0.19",` and `"vue-loader": "^15.9.7", "vue-template-compiler": "^2.6.12"` ):

Vue is processing everything inside of class as plain text and printing it like that straight to production, making our classes occupy much more precious bytes than they should:

```js
{staticClass:"\n          flex\n          items-center\n          font-bold\n          uppercase\n          text-h6\n          focus:outline-none\n        "} //          Yep, that's 10 individual space chars per class :(
```

This happens regardless of the configuration of whitespace preservation in `webpack.mix`:

```js
mix.vue({
  options: {
    compilerOptions: {
      whitespace: "condense", // doesn't condense classes :c
    },
  },
});
```

Therefore, having the option to disable this feature would solve cases where compilers interpret anything as plain text.
At least having the alternative of flexibility regarding this preference, in a similar fashion of quote styles or semicolons.

Thank you :)

**Comment by thorn0** [comment]
Thanks for the feedback. For now, the plan to improve this situation is to apply the new formatting only if:

* there already are line breaks in the `class` attribute OR
* any of the classes contains `--` as separator (Tailwind doesn't use it) AND the parser isn't `vue`

@adamwathan 

> If there are any questions I can answer or input I can provide that would be helpful please let me know

How does the plan above sound to you? The idea is to keep this new behavior at least for some developers that don't use Tailwind. I noticed that Tailwind doesn't use `--` whereas long class names in code examples provided by those who requested this feature often contain this separator. Any other ideas on how we can detect Tailwind?

> but if it's going to do anything I would really prefer if it doesn't do anything to the order of the classes

Prettier never reorders anything. It's one of the project's principles.

> Is this a big issue for people outside Tailwind projects?

The answer seemed to be no, but turns out there is also an issue with Vue (see the previous comment).

> maybe our team could prioritize working on a Tailwind plugin for Prettier

Unfortunately, Prettier currently lacks an API for such a plugin. We have an open issue about that: https://github.com/prettier/prettier/issues/10803 If your team could help us with moving that issue forward, it'd be awesome.

I know there already exist multiple Tailwind plugins for Prettier, but they use hacks instead of the official API and thus are fragile. E.g., such plugins [conflict](https://github.com/prettier/prettier/issues/10261) with each other. We need a real solution for this.

**Comment by royeden** [comment]
@thorn0 Any fix is welcome :D !

I'm not familiar with how prettier works under the hood, so I have to ask: is that approach easier than adding a config option for this feature? Or is it easier to validate it on a case by case relying on the parser and syntax?

I ask this because the issue could be affecting other users with other stacks, for example [this comment about htl](https://github.com/prettier/prettier/issues/10918#issuecomment-849789204) on the tread from @sabberworm .

Thank you for your response :)

**Comment by adamwathan** [comment]
@thorn0 Thanks for the update!

> For now, the plan to improve this situation is to apply the new formatting only if:
>
> - there already is a line break in the class attribute OR
> - any of the classes contains -- or __ as separator (Tailwind doesn't use them) AND the parser isn't vue

What do you think about simplifying this and _only_ applying the formatting if the class attribute already contains a line break? To me that feels a lot simpler than having a very specific rule about double dashes or double underscores, and makes it very easy for people to opt-in to the behavior if they prefer it, and would be very learnable/easy-to-understand what's triggering that formatting without having to read any documentation or anything to learn about the special `--` case.

**Comment by alexander-akait** [comment]
> What do you think about simplifying this and only applying the formatting if the class attribute already contains a line break? 

It is not heuristically, and kills the whole point of formatter

**Comment by adamwathan** [comment]
That is what was recommended by @thorn0  though, no?

> Thanks for the feedback. For now, the plan to improve this situation is to apply the new formatting only if:
> 
> - there already is a line break in the class attribute OR
> - any of the classes contains -- or __ as separator (Tailwind doesn't use them) AND the parser isn't vue

If their suggestion is against the philosophy of Prettier that is okay but it was not my suggestion, I was only suggesting to not base it on the presence of `--` or `__`.

**Comment by alexander-akait** [comment]
Maybe https://github.com/prettier/prettier/issues/10803 is real good solutions without any hacks...

**Comment by thorn0** [comment]
@adamwathan 

>  would be very learnable/easy-to-understand what's triggering that formatting without having to read any documentation or anything to learn about the special -- case

The idea is to make it "just work" for everybody without a need to read anything. My hypothesis is that people with these long class names with `--` and `__` never want long single-line `class` attributes.

@royeden

> is that approach easier than adding a config option

Not adding options is an important principle of Prettier. See https://prettier.io/docs/en/option-philosophy.html
We're always open to discussing heuristics though.

**Comment by iKlsR** [comment]
> My hypothesis is that people with these long class names with -- and __ never want long single-line class attributes.

I agree with Adam regarding _only_ considering the linebreak as that should be a good enough indicator of user intent. I personally don't want prettier to mess with my class strings at all. If I want to see everything I'll toggle wrapping in my editor. Thanks for your hard work on this project and hope I don't sound come off as entitled or rude, cheers.

**Comment by thorn0** [comment]
@iKlsR One of Prettier's principles is to leave as little as possible up to the user and another one is [formatting reversibility](https://prettier.io/docs/en/rationale.html#%EF%B8%8F-a-note-on-formatting-reversibility). If we can't fully stick to them in this case, we should at least try to minimize the digression. If we can prevent at least some nitpicky PR reviews like "please add a line break here for Prettier to reformat these classes", we should do this. BTW, do you use Tailwind? Do you use `--` and `__`?

**Comment by iKlsR** [comment]
@thorn0 Yes, I use a mixture of both across several projects, some older ones with BEM, some with Tailwind and in a few cases both where I haveuser themeable components.

**Comment by fabd** [comment]
Here is my feedback as well if that helps.

Same issue as OP, even a simpe loop in Vue becomes really big:

```
      <div class="mt-2 mb-4">
        <em class="inline-block mr-4">Intervals (days):</em>
        <span
          v-for="i in intervals"
          class="
            inline-block
            mr-2
            pt-1
            px-2
            pb-[0.15rem]
            rounded-md
            font-mono
            bg-[#d7e0b5]
            text-[#485f27]
            border-b border-[#aab38a]
          "
          >{{ i.days }}</span
        >
      </div>
```

Interestingly if we have few classes, the "smart formatting" in 2.3.0 doesn't trigger at all:

```
      <div class="mt-2 mb-4">
        <em class="inline-block mr-4">Intervals (days):</em>
        <span v-for="i in intervals" class="inline-block mr-2 pt-1 px-2">{{
          i.days
        }}</span>
      </div>
```

My proposed middle way solution is that you only apply that formatting *per line*, for any line that is too long, and then maybe as an added logic, any ocnsecutive "too long" lines, will be merged as "one too long line" for the smart formatting.

This means that I could do this, and Prettier shouldn't reformat the lines as they are not too long:

```
      <div class="mt-2 mb-4">
        <em class="inline-block mr-4">Intervals (days):</em>
        <span
          v-for="i in intervals"
          class="inline-block mr-2 pt-1 px-2 pb-[0.15rem]
            font-mono
            bg-[#d7e0b5] text-[#485f27] 
            border-b border-[#aab38a] rounded-md
          "
          >{{ i.days }}</span
        >
      </div>
```

My hunch is you'll find there are too many different preferences here, I feel that Prettier is reaching outside of its scope a little bit?

For example personally I have a habit, even when writing SCSS, of grouping layout together (display, padding, margin, ...). I also like to keep colors together (text; background). Then I will put borders together (border width, color, radius). So yo ucould say I group classes logically from a visual / designing standpoint and not from an abstract "by prefix" standpoint.

it may also be interesting to note, that Prettier used to let you insert newlines inside a class attribute already. It took me a while to notice this at my last job, and when I realized I could use newlines then I was able to group classes in a meaningful way, while at same time not taking too many lines either.

So the last example code I gave above worked. I would argue this is an area where "hands off" may be the better approach.

**Comment by fabd** [comment]
> @alexander-akait I guess it's a matter of preference then. For me, if I consider the following:

Yes, as a html/css focused frontend who enjoys the layout part, is is super super hard for me to work in a template like this.

From @adamwathan 

> What do you think about simplifying this and only applying the formatting if the class attribute already contains a line break?

That actually reverts how it worked prior to 2.3.0, in which you could group classes if you like by inteionally adding newlines. Prettier would let you "multiline" the class attribute and not touch it, which was great.

So this solution basically would give even less freedom then before, and force you to either have super long templates (many newlines from smart formatting), or have super long lines with no line breaks.

Imho prior to the change, it worked just fine. It give just enough freedom to html/css focused devs to multiline in a way that is meaningful to them, if they want to, and Prettier would leave those alone. In my experience working two years in a big Vue/NuxtJs  project, it was not an issue.

@iKlsR 

> I agree with Adam regarding only considering the linebreak as that should be a good enough indicator of user intent.

So in my case I disagree, it is not my intent when I have newlines in a class attribute; that they are reformatted. If anything, as I write a new component, having intentionally put newlines in there, means that I did it purposely.

@thorn0 

> Prettier never reorders anything. It's one of the project's principles.

I think that explains why the smart formatting can't work. To format class attribute in a meaningful way, especially for those devs that spend a lot of time editing templates (typically focused more on html/css as I am), they would have to be grouped in a meaningful way, that requires reordering the classes. So maybe it is best left to external plugins or utilities as Adam mentioned.

**Comment by rgilsimoes** [comment]
I was having this issue lately but found out the problem was related to Vetur extension and not Prettier extension. Rolled back Vetur to 0.33.1 and now it's fine! Does this make any sense?

**Comment by iKlsR** [comment]
@rgilsimoes No? That's the whole point of this long thread. :) Your vetur is using an older version of prettier as a dep. This change was introduced in 2.3.0, your vetur uses 2.2.1 under the hood.

https://github.com/vuejs/vetur/releases/tag/v0.34.0 *(Upgrade prettier to 2.3.0.)*
https://prettier.io/blog/2021/05/09/2.3.0.html#html

**Comment by rgilsimoes** [comment]
@iKlsR thanks' for pointing that out! I was mislead by the Prettier VS Code extension.
Using the latest prettier VSCode extension I don't experience that issue on other HTML files with tailwind classes.

Anyway I also don't agree with that style option! 
Neither with breaking tags like 
```
Some link</a
>
```

**Comment by surjithctly** [comment]
Can anyone tell me how to enable this feature? I'm using next.js and this single line feature is not working with VSCode. Do I need to do anything else?

**Comment by Shahaed** [comment]
> Can anyone tell me how to enable this feature? I'm using next.js and this single line feature is not working with VSCode. Do I need to do anything else? Seems its working with `class` but not `className`

It doesn't work in JSX, just in HTML. Look at #10663

**Comment by GustavoFenilli** [comment]
Can't figure it out on how to have formatting using vue + tailwind without having this problem, it seens the solution for now is just to disable veturs formatting and not formmating at all.

Does anyone have a solution to be able to format all but the template section for vue?

**Comment by adarsh4d** [comment]
Instead of disabling, downgrade/install the 0.33.1 version of vetur extension.

FYI prettier was upgraded to 2.3.0 on the 0.34.0 version.

**Comment by armenr** [comment]
I came here with the same problem as OP.

Currently, with windi/tailwind, my code format this way (even with line length of 80 set in prettier)

```
        <button
          type="button"
          class="inline-flex items-center px-4 py-2 border border-transparent text-base font-medium rounded-md shadow-sm text-black bg-white bg-opacity-75 hover:bg-opacity-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-gray-500"
          @click="router.back()"
        >
          {{ t("button.back") }}
        </button>
```

To me, this is a bit unwieldy. One of the beautiful things about a utility-first CSS approach is (among many other value-adds) a significant reduction in cognitive load (keeping track of various "global" style sheets + single-component `<style>` blocks.

To that end, I feel like being able to configure prettier to format it in the following way just makes the DX MUCH better, and also helps the html to read a lot more fluently...IMHO.

```
        <button
          type="button"
          class="inline-flex
            items-center
            px-4 py-2
            border
            border-transparent
            text-base
            font-medium
            rounded-md
            shadow-sm
            text-black
            bg-white
            bg-opacity-75
            hover:bg-opacity-50
            focus:outline-none
            focus:ring-2
            focus:ring-offset-2
            focus:ring-gray-500"
          @click="router.back()"
        >
```
So, I guess, just to double-check --> is there a way to configure prettier to always format it this way, currently?

**Comment by manojchandrashekar** [comment]
I really wish there was a "wrap class names at end of printWidth" as an alternative instead of each class name on a separate line.

**Comment by GustavoFenilli** [comment]
As sugested by @manojchandrashekar I think there should be a flag to make classes wrap on printWidth resulting on such, or at least a new width property like classWidth that would wrap only classes.

```
class="
    bg-opacity-75 bg-white border border-transparent focus:outline-none focus:ring-2
    focus:ring-gray-500 focus:ring-offset-2 font-medium hover:bg-opacity-50
    inline-flex items-center px-4 py-2 rounded-md shadow-sm text-base text-black
"
```

**Comment by klodoma** [comment]
In which version was this feature added? I am currently using `1.19.1`, tried to upgrade but  this formatting is messing up everything.

**Comment by alexcroox** [comment]
We are all stuck on v2.2.1 unfortunately

**Comment by saade** [comment]
Just found out that using [Headwind](https://github.com/heybourn/headwind) (mentioned by @adamwathan) to sort classes, it reorder my classes in one line =)

Sweet!

**Comment by alppe461** [comment]
Pr

**Comment by carlosmarioroldan** [comment]
I don't understand why anyone would add a feature that can mess up a lot of codebases without the possibility to opt-out.

Maybe an option in the config file where we can disable this annoying "feature"?

something like:

```json
{
  "formatHtmlClasses": false
}
```

**Comment by fabd** [comment]
Note to maintainers: I really didn't mean to make like a passive/aggressive comment by having one of my commits be cross-referenced here. I just copy/pasted the title of the issue in a commit as a note to self. I tried to remove this with edit and push --force. I'm not sure github will remove it but anyway, sorry for the distraction. It wasn't intended.

**Comment by Nness** [comment]
I would recommend to move on by using default formatter for HTML rather than prettier. 

e.g.
```
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "[html]": {
    "editor.defaultFormatter": "vscode.html-language-features"
  }
```

HTML already suffer in git commit, since minor change on the style or HTML tag will cause whole page to be restructured, better choose something consistent rather than opinionated. Any HTML formatter that can change the way it look from time to time is not suitable for Git.

You can choose to wait on the setting, and once they introduce they might introduce another breaking change, then you have to wait another setting.

**Comment by GustavoFenilli** [comment]
> I would recommend to move on by using default formatter for HTML rather than prettier.
> 
> e.g.
> 
> ```
>   "editor.defaultFormatter": "esbenp.prettier-vscode",
>   "[html]": {
>     "editor.defaultFormatter": "vscode.html-language-features"
>   }
> ```
> 
> HTML already suffer in git commit, since minor change on the style or HTML tag will cause whole page to be restructured, better choose something consistent rather than opinionated. Any HTML formatter that can change the way it look from time to time is not suitable for Git.
> 
> You can choose to wait on the setting, and once they introduce they might introduce another breaking change, then you have to wait another setting.

sadly, vue files are not html files, and vetur uses the new version of prettier.

**Comment by dgvai** [comment]
Still today, this problem is unsolved! Seriously man, is tailwind and prettier are enemy?

**Comment by rana-ruet** [comment]
Prettier 2.3 has this problem, You can use  Prettier 2.2 to avoid this hassle.

**Comment by alexcroox** [comment]
@adamwathan any chance of an update from the TW team on this? Do you guys still use prettier since the breaking change earlier this year, have you prevented it from updating itself in your tooling since? Are there plans to create some sort of prettier plugin to disable this behaviour? 

I'm asking as I'm sure this effects a lot of TW users and your team must also be impacted by this for your internal DX?

**Comment by adamwathan** [comment]
@alexcroox We are pinned to 2.2 internally. Not sure if a plugin can workaround this or not, unfortunately not familiar with authoring Prettier plugins.

**Comment by mreduar** [comment]
It would be great if this behavior could be avoided. I will stay on top of this issue.

**Comment by royeden** [comment]
Hey!

Here to report something else that's been coming up regarding class sorting (not sure if it should be brought up in a separate issue):

```html
<div class="grid bg-black grid-8 hover:bg-gray focus:bg-gray transition-colors duration-300 ease-in-out"></div>
```

is converted to

```html
<div
  class="
    grid
    bg-black
    grid-8
    hover:bg-gray
    focus:bg-gray
    transition-colors
    duration-300
    ease-in-out
  "
></div>
```

[Repro in playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAJgSwG4AIwA2AhgM4kC8AOiAOYBOm6uARjQLTPFgDWu9jbABy4AFhGxw6SVm3pEAnrgBmEMAFcS09nMUw6RKCUwxM0NpAIQ6JXOjX6TZgMwAGF7jik4bTFDYQ1GGoAPlQAeixsYJAAGhAIAAdHQ2RQIjo6CAB3AAV0hBJkECICLIVCuOZ9HjgYAGUiAFs4ABlfOGQlEpI4SuruWrqEojBfGmQ9NV6QOEbmOHR0BZaDGjUiGjgAMStGohgTKHGUIkCIWJARGEaCAHURYzgSYbA4OoLjHGN5IrAyC98PToMBy+hoe063WmACsSAAPOpjAhwACKagg8EhBB6cWG1kkRSuNwuCQYsFujBgImQghcuMyPVu+gSRVJT0kEguAEd0fBQYlCicSGwoHAFgsLnQ4DzMFLQRsIUgutjpj1GpgJnQpnEjEdkWiMR0kJNpjAiMwKegqcgAExxPRETAEMYAYQgjUVMxIAFYLho4AAVc2C5U4kDYKYASSgS1gdTADCSAEEY3UYPJkViegBfbNAA)

I've noticed that prettier groups classes by name, but it does **it only does it if these are next to each other**, the following html:

```html
<div class="bg-black grid grid-8 hover:bg-gray focus:bg-gray transition-colors duration-300 ease-in-out"></div>
```

works correctly and groups both `grid` and `grid-8` classes into a single line:

```html
<div
  class="
    bg-black
    grid grid-8
    hover:bg-gray
    focus:bg-gray
    transition-colors
    duration-300
    ease-in-out
  "
></div>

```

[Repro in playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAJgSwG4AIwA2AhgM4kC8AOiAEYDmAtDcWANa50BOm6H36DABy4AFhGxxOSegy5EAnrgBmEMAFcS0xnMUxORKCUwxM0BpAIROJXOjX6TZgMwAGF7jik4DTFAYQ1GGoAPlQAeixsYJAAGhAIAAdHQ2RQIk5OCAB3AAV0hBJkECICLIVCuJp9NjgYAGUiAFs4ABlfOGQlEpI4SurWWrqEojBfOmQ9NV6QOEaaOHR0BZaDOjUiOjgAMStGohgTKHGUIkCIWJARGEaCAHURYzgSYbA4OoLjHGN5IrAyC98PU4MBy+joe063WmACsSAAPOpjAhwACKagg8EhBB6cWG1kkRSuNwuCW4sFuPBgImQghcuMyPVu+gSRVJT0kEguAEd0fBQYlCicSAwoHAFgsLpw4DzMFLQRsIUgutjpj1GpgJpwpnEjEdkWiMR0kJNpjAiDQKegqcgAExxPRETAEMYAYQgjUVMxIAFYLho4AAVc2C5U4kDYKYASSgS1gdTA3CSAEEY3UYPJkViegBfbNAA)

**Comment by yoyo930021** [comment]
> > I would recommend to move on by using default formatter for HTML rather than prettier.
> > e.g.
> > ```
> >   "editor.defaultFormatter": "esbenp.prettier-vscode",
> >   "[html]": {
> >     "editor.defaultFormatter": "vscode.html-language-features"
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > HTML already suffer in git commit, since minor change on the style or HTML tag will cause whole page to be restructured, better choose something consistent rather than opinionated. Any HTML formatter that can change the way it look from time to time is not suitable for Git.
> > You can choose to wait on the setting, and once they introduce they might introduce another breaking change, then you have to wait another setting.
> 
> sadly, vue files are not html files, and vetur uses the new version of prettier.

https://github.com/vuejs/vetur/issues/3156#issuecomment-935821646
You can use another formatted in vetur.

**Comment by PatricNox** [comment]
Working in tailwind and being forced to this rule is just completely horrid

**Comment by adamwathan** [comment]
Just to share another interesting opinion on this, @gaearon had this to say on Twitter recently:

https://twitter.com/dan_abramov/status/1452328297036193794?s=21

> (Tailwind) is very terse. you can have
> 
> `<div className="... 100 chars ...">`
> 
> this SEEMS bad at first, but it’s like “styles are collapsed by default”. unlike styled-comps or inline CSS, it prevents the tree from exploding despite colocation. and really long things can be components.
> 
> so it feels like inline styles (which i generally enjoy using a lot because i don’t have to come up with names for every little thing), but without spreading my tree too thin vertically, and without the perf cost of inline styles.

I think it's an important benefit as well. Ultimately of course it's up to the wonderful folks doing the work to maintain the tool ❤️ but if outside feedback is to be considered, my preference would be to just keep things simple the way they behaved before, where class lists aren't split.

**Comment by mnr73** [comment]
for solve this problem install this extension

https://marketplace.visualstudio.com/items?itemName=Wscats.vue

and enable auto form
![2021-10-30_10-00-26](https://user-images.githubusercontent.com/23061025/139523014-7a78df9b-794c-4508-9913-c20fcd28dcf0.png)

for disable eslint warning you set "on" no-console in .eslintrc.js
![2021-10-30_10-02-25](https://user-images.githubusercontent.com/23061025/139523061-e57d19ea-d66d-4145-8d6c-0181c95977ba.png)

**Comment by lkj4** [comment]
I think @GustavoFenilli [solution ](https://github.com/prettier/prettier/issues/10918#issuecomment-886125119) (just wrap neatly at printWidth and indent respectively the next line) is the best because...

- it works also with integrated code blocls, e.g. in Svelte, btw when will this feature be turned on for svelte files?
- keeps the code compact
- it is generic, not everybody is using tailwind, there are still other around, e.g.,  tachyons, unocss, windi, whatever

only disadvantage is that one mentioned, git diffs might look a bit crowded but this is better than having blown up 100 lines components because of one class per line and no ripped apart code blocks.

i agree with @thorn0 that this option should just work without any options out of the box or some insider hacks which the tailwind maintainer suggested.

q how do we proceed, democratic process or let @thorn0 as the maintainer just decide?

**Comment by Shahaed** [comment]
Sorry to rehash my earlier comment, but part of Prettier's philosophy for new line objects is if there's a new line after the object it will expand entires to new lines, and if not it will keep them in line (up to line length) -- [see here](https://prettier.io/docs/en/rationale.html#multi-line-objects).

It could make sense to reuse this logic for CSS class attributes. So if the user has put a new line after then format the class attribute string. Else leave it as a single line and remove extra spaces or whatever else. This should be consistent with the Prettier philosophy of no added options too.

Some users of Tailwind prefer newlines (my team and the team at [Netlify](https://www.netlify.com/blog/2021/03/23/from-semantic-css-to-tailwind-refactoring-the-netlify-ui-codebase/#readability) since it makes reading the CSS easier ESPECIALLY for users that don't use editor text wrapping.

**Comment by fabd** [comment]
> just wrap neatly at printWidth and indent respectively the next line

I don’t think this is a good default setting, because it’s easy to get lots of indentation in html template even for a single component. So the printWidth layout would look terrible if breaking a bunch of css classes in some html that’s already indented half way through a line.

**Comment by iKlsR** [comment]
At this point in time, I think *a lot* of people would be very happy and able to move from frozen versions of packages etc if this were just reverted or at least able to be disabled, my 2c and again being as respectful as possible to the developers on their stance and effort as while this is welcome to some, the majority clearly says otherwise.

Funnily enough tailwind wasn't even the sole reason I came to this, it just messed up all my projects and workflow, it's a very strong opinion to enforce out of the blue and in my use case with my eyesight, monitor and font size, several vertical lines of css or classes that I will rarely interact with is a lot of screen real estate and I'm sure it's the same for others as well.

**Comment by dillingham** [comment]
@iKlsR yes

**Comment by dillingham** [comment]
that class stack is not "prettier'

**Comment by iamnbutler** [comment]
FWIW this behavior will likely make me move away from prettier as well – It's just a little too shocking to have a massive stack of CSS classes explode out on format. I do appreciate the principle of no options, I just think this is an edge it runs up against. It is unfortunate because otherwise, I do enjoy using prettier.

**Comment by klodoma** [comment]
This discussion is becoming slowly pathetic. Several options were proposed nothing happens. Because of this we're stuck with version `2.2.1`. 

As much as _"Prettier is built on the principle of being opinionated about code formatting"_, just make another option to disable this so called "feature". 

I cannot understand why **such an option would violate the previous principle**, otherwise you would have hard-coded `printWidth` as well. Or?

**Comment by Christopher-Hayes** [comment]
I appreciate the idea, but as a Tailwind user, this format is just not usable when 1-3 DOM elements fill the entire screen.

**Comment by adamwathan** [comment]
Here's my personal wishlist for how this would work in a perfect world:

1. Keeps classes on one really long line if the class string contains no newlines.
2. Removes unnecessary whitespace between class names.
3. Adds new lines after `class="` and before the closing `"` if the class string contains any newlines.
4. Doesn't add extra newlines inside the class string ever, except for the starting and ending ones, which are only added if there is already a newline in the class string.
5. Keeps `class="` on the same line as the opening tag if it's the only attribute, even when the class string is broken across multiple lines.

Examples:

```html
<!-- Input -->
<div class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
  Hello world
</div>

<!-- Output -->
<div class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
  Hello world
</div>
```

```html
<!-- Input -->
<div class="mt-8   w-full  bg-indigo-600    border">
  Hello world
</div>

<!-- Output -->
<div class="mt-8 w-full bg-indigo-600 border">
  Hello world
</div>
```

```html
<!-- Input -->
<div class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
hover:bg-indigo-700
    focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
    Hello world
</div>

<!-- Output -->
<div class="
  mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
  hover:bg-indigo-700
  focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500
">
    Hello world
</div>

```html
<!-- Input -->
<div some-attribute some-other-attribute class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
hover:bg-indigo-700
    focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
    Hello world
</div>

<!-- Output -->
<div
  some-attribute
  some-other-attribute
  class="
    mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
    hover:bg-indigo-700
    focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500
  "
>
    Hello world
</div>
```

Is anyone interested in getting a PR started for this maybe? This is the behavior I personally am able to get behind the most, and maybe having a PR to support can help push this forward. Of course if the maintainers don't believe this behavior aligns with the spirit of the project that's okay too, but hoping we can at least start somewhere.
    
I'm probably missing some hard-to-solve cases here so happy to discuss if helpful.

**Comment by GustavoFenilli** [comment]
> Here's my personal wishlist for how this would work in a perfect world:
> 
> 1. Keeps classes on one really long line if the class string contains no newlines.
> 2. Removes unnecessary whitespace between class names.
> 3. Adds new lines after `class="` and before the closing `"` if the class string contains any newlines.
> 4. Doesn't add extra newlines inside the class string ever, except for the starting and ending ones, which are only added if there is already a newline in the class string.
> 5. Keeps `class="` on the same line as the opening tag if it's the only attribute, even when the class string is broken across multiple lines.
> 
> Examples:
> 
> ```
> 
> <div class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
>   Hello world
> </div>
> 
> 
> <div class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
>   Hello world
> </div>
> ```
> 
> ```
> 
> <div class="mt-8   w-full  bg-indigo-600    border">
>   Hello world
> </div>
> 
> 
> <div class="mt-8 w-full bg-indigo-600 border">
>   Hello world
> </div>
> ```
> 
> ```
> 
> <div class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
> hover:bg-indigo-700
>     focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
>     Hello world
> </div>
> 
> 
> <div class="
>   mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
>   hover:bg-indigo-700
>   focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500
> ">
>     Hello world
> </div>
> ```
> 
> ```
> 
> <div some-attribute some-other-attribute class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
> hover:bg-indigo-700
>     focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
>     Hello world
> </div>
> 
> 
> <div
>   some-attribute
>   some-other-attribute
>   class="
>     mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white
>     hover:bg-indigo-700
>     focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500
>   "
> >
>     Hello world
> </div>
> ```
> 
> Is anyone interested in getting a PR started for this maybe? This is the behavior I personally am able to get behind the most, and maybe having a PR to support can help push this forward. Of course if the maintainers don't believe this behavior aligns with the spirit of the project that's okay too, but hoping we can at least start somewhere.
> 
> I'm probably missing some hard-to-solve cases here so happy to discuss if helpful.

I think this is a good solution, with this you could also use another tool to agregate similar classes with a line break so it would wrap around in a way that makes sense.

```html
<div class="
  mt-8 p-8
  bg-indigo-600 hover:bg-indigo-700
  border border-transparent
  font-medium text-white
">
  Hello world
</div>
```

**Comment by mreduar** [comment]
> Here's my personal wishlist for how this would work in a perfect world:

Could you please make a gist of how to configure it this way?

**Comment by alexander-akait** [comment]
@adamwathan 

> Keeps classes on one really long line if the class string contains no newlines.

I'm afraid this is not a good idea, we must not rely on user input, we should rely on string length, as you can see some examples are very long and it is not readable...

**Comment by adamwathan** [comment]
@alexander-akait  Hey Alexander!

> I'm afraid this is not a good idea, we must not rely on user input, we should rely on string length, as you can see some examples are very long and it is not readable...

This is how Prettier behaved prior to this recent change though, it never added newlines into classes for you.

It did this:

```html
<!-- Input -->
<div class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
  Hello world
</div>

<!-- Output -->
<div
  class="mt-8 w-full bg-indigo-600 border border-transparent rounded-md py-3 px-8 flex items-center justify-center text-base font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
>
  Hello world
</div>
```

...so the class list stayed as one long string (which I think is good), but it moved the actual attribute to its own line (which I personally don't think is useful).

The major complaint here (and the reason this issue was opened) is that generally people (myself included) seem to _prefer_ the really long class names staying on one line, because breaking it up onto multiple lines makes templates very difficult to work with because every HTML element can take up like 12 lines in your editor.

The comment here about "styles are collapsed by default" is relevant:

https://github.com/prettier/prettier/issues/10918#issuecomment-955096809

For our HTML stuff at Tailwind Labs we are pinned to Prettier 2.2 and actually set `printWidth` to `10000` to get the behavior we want where the class attribute stays on the same line, haha... Unfortunately this is a bad setting when working in JSX though, so there we are just pinned to 2.2 and deal with `class` being put on its own line unnecessarily.

**Comment by GustavoFenilli** [comment]
> @adamwathan
> 
> > Keeps classes on one really long line if the class string contains no newlines.
> 
> I'm afraid this is not a good idea, we must not rely on user input, we should rely on string length, as you can see some examples are very long and it is not readable...

If user input is not a desired behaviour, you could instead break the lines based on print width, so instead of having 20 lines it would be like 4, that is much more manageable, unless there is already a break so we know that the user wants the behaviour of breaking those lines, falling down to @adamwathan ideas.

**Comment by alexander-akait** [comment]
@adamwathan So the main problem is a lot of lines inside the `class` attribute that interfere with readability?

**Comment by luukdv** [comment]
@alexander-akait Exactly! See https://github.com/prettier/prettier/issues/10918#issuecomment-845708671 for an example on how it's detrimental for readability.

**Comment by adamwathan** [comment]
@alexander-akait Yeah correct, I think the community finds the code on the right much more manageable than the code on the left:

![image](https://user-images.githubusercontent.com/4323180/142220687-e957c5db-a165-41fd-aad9-a5ab67ac8bee.png)

**Comment by alexander-akait** [comment]
@adamwathan Yep, sounds reasonable

@thorn0 @sosukesuzuki @fisker Thoughts?

**Comment by fabd** [comment]
> (...) is that generally people (myself included) seem to prefer the really long class names staying on one line

I agree and it's worth considering most code editors have "word wrap" feature, so editing a long line of class attributes isn't too bothersome.

Interestingly ; I never even paid attention before ; but both VSCode and Sublime Text even re-indent after wordwrap even through it is a _single_ line:

![Screenshot 2021-11-17 at 17 03 19](https://user-images.githubusercontent.com/169391/142236236-05b20e07-ddea-400f-a9f0-66abcc6b9ca6.png)


> This is how Prettier behaved prior to this recent change though, it never added newlines into classes for you.

There are several updates if that helps related to the change of behaviour:

- I htink it started in #7555 [Prettier update 2.0.0](https://prettier.io/blog/2020/03/21/2.0.0.html#format-value-of-html-class-attribute-7555httpsgithubcomprettierprettierpull7555-by-fiskerhttpsgithubcomfisker) where it removed extra spaces (nice), but also started tinkering with (usually) intended newlines in the class attribute

- then there was further update(s) that did the smart formatting, if not mistaken

**Comment by fisker** [comment]
I think we can just revet #7555 and #7865 to make things simple.

**Comment by lkj4** [comment]
prettier's philosophy is not about respecting deliberately set newlines, this is too much mental drain. just let prettier do the change without being directed by the users or some newlines within the class attribute. I want opinionated defaults.

to the comparisons between one class per line vs all in one line, this comparisons are jarring, there is still some middle ground between these options. 

I think the maintainer should decide on its own. prettier is for a reason popular, because this guy made smart design choices in the past.

**Comment by lkj4** [comment]
> 
> 
> @alexander-akait Yeah correct, I think the community finds the code on the right much more manageable than the code on the left:
> 

you should compare to wrapped at print width not all in one line, then people would vote differently

**Comment by dillingham** [comment]
> prettier's philosophy is not about respecting deliberately set newlines, this is too much mental drain. just let prettier do the change without being directed by the users or some newlines within the class attribute. I want opinionated defaults.
> 
> 
> 
> to the comparisons between one class per line vs all in one line, this comparisons are jarring, there is still some middle ground between these options. 
> 
> 
> 
> I think the maintainer should decide on its own. prettier is for a reason popular, because this guy made smart design choices in the past.

30 lines of classes for an element is anything but "pretty"

**Comment by klodoma** [comment]
Why not just sort the classes and wrap them at `printWidth`? 
I think most people are bothered by the multiple `EOLs` that are introduced and not by the fact that the classes are wrapped in 2 or 3 lines. 

I think these examples would fit well within 2 lines. 

> @alexander-akait Yeah correct, I think the community finds the code on the right much more manageable than the code on the left:
> 
> ![image](https://user-images.githubusercontent.com/4323180/142220687-e957c5db-a165-41fd-aad9-a5ab67ac8bee.png)

**Comment by leodr** [comment]
> Why not just sort the classes and wrap them at `printWidth`? I think most people are bothered by the multiple `EOLs` that are introduced and not by the fact that the classes are wrapped in 2 or 3 lines.
> 
> I think these examples would fit well within 2 lines.
> 
> > @alexander-akait Yeah correct, I think the community finds the code on the right much more manageable than the code on the left:
> > ![image](https://user-images.githubusercontent.com/4323180/142220687-e957c5db-a165-41fd-aad9-a5ab67ac8bee.png)

I don’t think sorting them in any automated, non-framework-specific way makes sense, especially when using atomic CSS like Tailwind. There you would much rather group related classes like `p-4` and `m-4` than having them alphabetically ordered.

**Comment by lkj4** [comment]
Guys, guys, let's make small steps: Talking about reordering and word-wrapping is too much at once. These are two different features with entirely different requirements. I'd suggest to find an agreement, first, on the word-wrapping and once we are there we can think about reordering (if at all since reordering is something not really in the realm of prettier), ok?

For the sake of simplification, we have currently two parties in this thread: 

🎄 The Xmas-Tree Hardliners (one class per line)
- good: nice git diff[1]
- con: ultra bloated components looking like xmas all the year with gazillions LOC, but good for your perf review haha 

🥞 The Pancake Party (just wrap at print width and intend next line accordingly)
- good: keeps code nice and compact

Did I miss any party? If not, let's start to find an agreement just on this topic before we move on to reordering, ok?

@thorn0 I think you got fed up with this thread, but let's at least try to get closer to some solution. The current state of matters is really hard to work with. Thanks!

[2] i find this pro really negligible, I never really read git diffs of some css classes

**Comment by bodograumann** [comment]
The third option is (as a first simple step at least) to revert and go back to leaving class strings alone, like @fisker suggested.
I also think if voting is to be done, it should be using reaction emojis, not replies.
Maybe :grinning: for revert, :tada: for leaving as it is currently (christmas tree) and :rocket: for adding breaks at print width (pancake party).
Though there is no probably not much use in voting because people are really unhappy with the current state :tada: and breaking at print width :rocket: was rejected by the maintainers.

**Comment by adamwathan** [comment]
It's been reverted 👍🏻 just need to wait for a new release.

**Comment by nicooprat** [comment]
Cf https://github.com/prettier/prettier/pull/11827

**Comment by fabd** [comment]
Thanks for the update. Definitely more usable.

I would still like to see a way to leave intentional newli es.

**Comment by hivokas** [comment]
It has been released: https://github.com/prettier/prettier/releases/tag/2.5.0

**Comment by alexander-akait** [comment]
Let's close, feel free to feedback

**Comment by lkj4** [comment]
> Let's close, feel free to feedback

IDK, the latest release is a good baseline which works for all us, nobody will die and it's better than the Xmas tree. But are we there yet?

Closing the issue would be feel a bit like giving up. I mean the future of CSS will be for a significant part about utility CSS. And maybe the solution is that the Tailwind team or new teams need to come up with a more terse API which fit most of the time into one line (Tachyons was better with this btw). Tailwind is in some areas terse but in many parts not much shorter than the og style equivalent.

Or, the Tailwind team just forks prettier and focuses more on IDE tooling than some ui toolkits. Former is way more crucial than some random ui components. /cc @adamwathan 

Or, there is a solution in better formatting which works for all of us, can be implemented by the og prettier team, without fragmentation of the community. I don't want to get again into the topic and repeat my view like a broken record. We need instead another take how to discuss with a proper moderation. Maybe we start with listing requirements of everyone before we sketch another solution. Some love git diffs, some have huge code blocks within template string, some need ordering but not alphabetically and so on. 

Utility CSS is too important that we just do nothing or play laissez faire. The current solution is ok-ish but far from what I would expect from prettier.

**Comment by jamisuomalainen** [comment]
> Closing the issue would be feel a bit like giving up.

I don't think anyone's giving up; you can find the follow-up discussion in https://github.com/prettier/prettier/issues/7863 as mentioned in the [2.5.0 release notes](https://prettier.io/blog/2021/11/25/2.5.0.html#collapse-html-class-attributes-onto-one-line-11827httpsgithubcomprettierprettierpull11827-by-jlongsterhttpsgithubcomjlongster).

**Comment by dillingham** [comment]
> Or, the Tailwind team just forks prettier and focuses more on IDE tooling than some ui toolkits. Former is way more crucial than another set of random ui components. /cc @adamwathan

You seem passionate about it, you could fork it also? Those UI components pay for the many full time devs on team tailwind and bring me more value that re ordering css classes. "Crucial" is subjective

**Comment by n4ks** [comment]
I was very upset by this change in version 2.5.0. It is much more convenient for me to work with classes when they are arranged in a column, and not in one row. Maybe you can add a setting for this instead of changing the default behavior? It would be very convenient and would allow everyone to work as it is convenient for him.

**Comment by bodograumann** [comment]
@fisker The other ticket #7863 is still marked as "locked as resolved and limited to collaborators."

**Comment by Scotchester** [comment]
Continuing the discussion here since the issue referenced in the release notes is still locked.

> > Here's my personal wishlist for how this would work in a perfect world:
> > 
> > 1. Keeps classes on one really long line if the class string contains no newlines.
> > 2. Removes unnecessary whitespace between class names.
> > 3. Adds new lines after `class="` and before the closing `"` if the class string contains any newlines.
> > 4. Doesn't add extra newlines inside the class string ever, except for the starting and ending ones, which are only added if there is already a newline in the class string.
> > 5. Keeps `class="` on the same line as the opening tag if it's the only attribute, even when the class string is broken across multiple lines.
> >
> > …
> >
> > Is anyone interested in getting a PR started for this maybe? This is the behavior I personally am able to get behind the most, and maybe having a PR to support can help push this forward. Of course if the maintainers don't believe this behavior aligns with the spirit of the project that's okay too, but hoping we can at least start somewhere.
> > I'm probably missing some hard-to-solve cases here so happy to discuss if helpful.
> 
> I think this is a good solution, with this you could also use another tool to agregate similar classes with a line break so it would wrap around in a way that makes sense.
> 
> ```
> <div class="
>   mt-8 p-8
>   bg-indigo-600 hover:bg-indigo-700
>   border border-transparent
>   font-medium text-white
> ">
>   Hello world
> </div>
> ```

I agree that this would be a really elegant solution that bridges the gap between the two extremes of one giant line and one line per class, neither of which are appealing to me as a Tailwind user.
